### ISSUE TITLE
Factorization twice as slow on 0.6 than 0.5

### ISSUE THREAD

**Issue by jebej** [JuliaLang/julia#21624]
```julia
A = rand(40,40);
@benchmark sqrtm($A)
```

0.6: (55c97fb)
```
BenchmarkTools.Trial:
  memory estimate:  269.69 KiB
  allocs estimate:  26
  --------------
  minimum time:     3.855 ms (0.00% GC)
  median time:      3.999 ms (0.00% GC)
  mean time:        4.105 ms (0.53% GC)
  maximum time:     12.227 ms (15.49% GC)
  --------------
  samples:          1216
  evals/sample:     1
```

0.5.1:
```julia
BenchmarkTools.Trial:
  memory estimate:  269.69 KiB
  allocs estimate:  26
  --------------
  minimum time:     1.731 ms (0.00% GC)
  median time:      1.795 ms (0.00% GC)
  mean time:        1.868 ms (1.13% GC)
  maximum time:     9.239 ms (0.00% GC)
  --------------
  samples:          2653
  evals/sample:     1
```

**Comment by fredrikekre** [comment]
Its due to these regression as noted in #20993:

| `["linalg","factorization",("schur","Matrix",1024)]` | 2.10 (45%) :x: | 1.00 (1%)  |
| `["linalg","factorization",("schur","Matrix",256)]` | 2.36 (45%) :x: | 1.00 (1%)  |
| `["linalg","factorization",("schurfact","Matrix",1024)]` | 2.09 (45%) :x: | 1.00 (1%)  |
| `["linalg","factorization",("schurfact","Matrix",256)]` | 2.36 (45%) :x: | 1.00 (1%)  |

**Comment by jebej** [comment]
Right. Do we have an idea what caused those? Isn't that all handled by Lapack? A factor of two seems high.

**Comment by fredrikekre** [comment]
Code path have not changed, and not the lapack wrapper either. But the factor 2 are there:

```julia
julia> A = rand(256, 256); cA = complex(A);

julia> @btime LinAlg.LAPACK.gees!('V', copy($cA)); # 0.6
  200.070 ms (11 allocations: 2.14 MiB)

julia> @btime LinAlg.LAPACK.gees!('V', copy($cA)); # 0.5.1
  90.971 ms (11 allocations: 2.14 MiB)
```

**Comment by andreasnoack** [comment]
Same version of OpenBLAS?

**Comment by jebej** [comment]
Nothing obvious shows up when profiling.
0.6:
```
194 .\REPL[2]:1; macro expansion
 3   .\array.jl:126; copy(::Array{Complex{Float64},2})
 191 .\linalg\lapack.jl:5586; gees!(::Char, ::Array{Complex...
```
0.5.1:
```
66 .\REPL[27]:1; macro expansion;
 65 .\linalg\lapack.jl:5584; gees!(::Char, ::Array{Complex{Float64},2})
 1  .\linalg\lapack.jl:5596; gees!(::Char, ::Array{Complex{Float64},2})
```

**Comment by jebej** [comment]
Would it be an issue with the `ccall` then?

**Comment by KristofferC** [comment]
These calls should be long enough for `ccall` overhead not to matter. Perhaps some difference in build flags used?

**Comment by jebej** [comment]
I can't compile julia from scratch (Windows) but is there an archive of old nightlies to see if I can identify when the problem started occurring? I think this is a pretty big issue that really should be fixed...

**Comment by tkelman** [comment]
julianightlies.s3.amazonaws.com goes back about a month

**Comment by stevengj** [comment]
I thought `sqrtm` was supposed to be faster in 0.6 from #20214.

**Comment by jebej** [comment]
According to [this report](https://github.com/JuliaCI/BaseBenchmarkReports/blob/aee96ad2822bef035f2a1d771d3f5da046f0ece3/9225708_vs_6445c82/report.md) the regressions date from before March 8th.

**Comment by stevengj** [comment]
@jebej, #20214 was merged on Feb 3; can you check whether it is the source of the regression?

**Comment by jebej** [comment]
I believe that PR seems was only for `sqrtm` of Upper Triangular matrices. The regressions in the OP was sqrtm for generic matrices, which uses `schurfact`, and thus `gees`.

**Comment by stevengj** [comment]
@jebej, `schurfact` is used to *get* an upper triangular matrix from an arbitrary matrix, at which point the `sqrtm` code calls `sqrtm(UpperTriangular(SchurF[:T]))`.  So, #20214 still seems relevant.

**Comment by stevengj** [comment]
Oh, but I see that you are specifically benchmarking `gees` above and still seeing a slowdown.

**Comment by jebej** [comment]
You're right, my mistake. Yes, the slowdown seems to be with `gees` in that case, and there is a similar one with `geevx`.

**Comment by stevengj** [comment]
I can't reproduce the `gees!` slowdown on MacOS. 
```
julia> A = rand(256, 256); cA = complex(A);

julia> @btime LinAlg.LAPACK.gees!('V', copy($cA));  # 0.5.1
  257.194 ms (11 allocations: 2.14 MiB)

julia> @btime LinAlg.LAPACK.gees!('V', copy($cA)); # 0.6
  251.435 ms (11 allocations: 2.14 MiB)
```

**Comment by stevengj** [comment]
Could it be a threading issue on your machine?  Try `BLAS.set_num_threads(1)`.

**Comment by jebej** [comment]
Is the optimization level the same between macOS and Linux? I vaguely recall a change from -O3 to -O2 for some dependencies, but I can't find the issue.

**Comment by jebej** [comment]
It's not just my machine, the slowdown is detected by [nanosoldier](https://github.com/JuliaLang/julia/pull/20993).

**Comment by fredrikekre** [comment]
> I can't reproduce the `gees!` slowdown on my Mac

I just tried on my Mac as well and I still see a slowdown with a factor 2...

**Comment by andreasnoack** [comment]
It might be useful if people would report `versioninfo()` as well. Maybe we can narrow down if this only happens on some architectures.

**Comment by fredrikekre** [comment]
I can see a slowdown of 2 (or more) on the following 3 setups (comparing `0.5.1` and the `release-0.6` branch):

```
julia> versioninfo()
Julia Version 0.6.0-pre.beta.460
Commit 63ae651* (2017-05-02 12:18 UTC)
Platform Info:
 OS: Linux (x86_64-linux-gnu)
 CPU: Intel(R) Core(TM) i5-7600K CPU @ 3.80GHz
 WORD_SIZE: 64
 BLAS: libopenblas (USE64BITINT NO_AFFINITY HASWELL)
 LAPACK: libopenblas64_
 LIBM: libopenlibm
 LLVM: libLLVM-3.9.1 (ORCJIT, broadwell)
```

```
julia> versioninfo()
Julia Version 0.6.0-pre.beta.460
Commit 63ae651d91* (2017-05-02 12:18 UTC)
Platform Info:
 OS: macOS (x86_64-apple-darwin16.5.0)
 CPU: Intel(R) Core(TM) i5-4258U CPU @ 2.40GHz
 WORD_SIZE: 64
 BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
 LAPACK: libopenblas64_
 LIBM: libopenlibm
 LLVM: libLLVM-3.9.1 (ORCJIT, haswell)
```

```
julia> versioninfo()
Julia Version 0.6.0-pre.beta.460
Commit 63ae651 (2017-05-02 12:18 UTC)
Platform Info:
 OS: Linux (x86_64-linux-gnu)
 CPU: Intel(R) Core(TM) i7-6600U CPU @ 2.60GHz
 WORD_SIZE: 64
 BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
 LAPACK: libopenblas64_
 LIBM: libopenlibm
 LLVM: libLLVM-3.9.1 (ORCJIT, skylake)
```

**Comment by fredrikekre** [comment]
I am so confused. With this:

```julia
using BenchmarkTools
function benchmark()
    versioninfo()
    A = rand(256, 256)
    cA = complex(A)
    for i in 1:3
        println(minimum(@benchmark LinAlg.LAPACK.gees!('V', copy($cA))))
    end
end
benchmark()
```

I get the following on **the same commit** the only difference being that one is from the `release-0.6` branch and the other is on the `master` branch.

609b3d12c75bdf99d98a29ab3db856a307253e74 checked out from `master`:
```
julia> benchmark()
Julia Version 0.6.0-pre.beta.459
Commit 609b3d1* (2017-05-02 04:05 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i7-6600U CPU @ 2.60GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, skylake)
TrialEstimate(252.737 ms)
TrialEstimate(247.091 ms)
TrialEstimate(218.330 ms)
```

609b3d12c75bdf99d98a29ab3db856a307253e74 checked out from `release-0.6`:
```
julia> benchmark()
Julia Version 0.6.0-pre.beta.459
Commit 609b3d1 (2017-05-02 04:05 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i7-6600U CPU @ 2.60GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, skylake)
VERSION
TrialEstimate(513.989 ms)
TrialEstimate(514.151 ms)
TrialEstimate(504.442 ms)
```

**Comment by fredrikekre** [comment]
So it turns out that my directory for the `master` branch has a fast version of `libopenblas`, and compiling from scratch (which I did for the `release-0.6` version) compiles a slow version. I just tried compiling everything from scratch and that results in a slow version.

My conclusion is that some build flags must have changed, but did not trigger a recompile of the library in my `master` directory (which I have been updating continuously).
